File: src/04.stack/07.20.ts
export function isValid(s: string): boolean {
    const stack: string[] = [];
    const map = {
        "}": "{",
        ")": "(",
        "]": "[",
    };

    for (let i = 0, n = s.length; i < n; i++) {
        if (s[i] === "[" || s[i] === "(" || s[i] === "{") {
            stack.push(s[i]);
        } else {
            if (stack.length && stack[stack.length - 1] === map[s[i]]) {
                stack.pop();
            } else {
                return false;
            }
        }
    }

    return !stack.length;
}

File: src/04.stack/07.20.go
package stack

func IsValid(s string) bool {
	stack := []rune{}
	obj := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}

	for _, v := range s {
		if v == '(' || v == '[' || v == '{' {
			stack = append(stack, v)
		} else {
			if stack[len(stack)-1] == obj[v] {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}

	return len(stack) == 0
}
